File: leetcode/flatten_binary_tree_to_linked_list/FlattenBinaryTreeToLinkedList.cpp
#include <iostream> 
#include <stack> 

using namespace std; 

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    void flatten(TreeNode *root) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        if (!root) {
            return; 
        }
        
        stack<TreeNode *> s; 
        s.push(root);  
        TreeNode * curr = NULL; 
        TreeNode * prev = NULL; 
        
        while (curr != NULL || s.size() != 0) {
            if (curr != NULL) {
                if (curr->right != NULL) {
                    s.push(curr->right); 
                }
                
                prev = curr; 
                TreeNode * temp = curr->left; 
                curr->right = curr->left; 
                curr->left = NULL; 
                curr = temp; 
            }
            else {
                curr = s.top(); 
                s.pop();
                
                if (prev != NULL) {
                    prev->right = curr; 
                } 
            }
            
        }
    }
    
    void inorder(TreeNode * root) {
        if (!root) {
            return; 
        }
        
        cout << root->val << " "; 
        inorder(root->left);
        inorder(root->right);  
    } 
};

int main() {
    Solution sln; 
    TreeNode * root = new TreeNode(1);
    /*
    root->left = new TreeNode(2); 
    root->left->left = new TreeNode(3); 
    */
    root->right = new TreeNode(2); 
     
    sln.flatten(root); 
    
    sln.inorder(root); 
    return 0; 
}

File: lintcode/_flatten_binary_tree_to_linked_list.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root: a TreeNode, the root of the binary tree
     * @return: nothing
     * 
     * Iterative pre-order tree traversal. 
     * O(n)
     */
    public void flatten(TreeNode root) {
        TreeNode prev = null; 
        TreeNode curr = root; 
        Stack<TreeNode> stack = new Stack<>(); 
        
        while (curr != null || stack.size() != 0) {
            if (curr != null) {
                if (curr.right != null) {
                    stack.push(curr.right);
                }
                
                curr.right = curr.left; 
                curr.left = null; 
                prev = curr; 
                curr = curr.right; 
            } else {
                curr = stack.pop();
                prev.right = curr; 
            }
        }
    }
}
